src/mds/Locker.cc

```diff
@@ -3877,7 +3877,7 @@ bool Locker::_do_cap_update(CInode *in, Capability *cap,
 
   Session *session = mds->get_session(m);
   if (session->check_access(in, MAY_WRITE,
```

Comment:
As we discussed in the meeting today:

* I don't think this check belongs in `_do_cap_update`. We check authorization in cap issuance, not during update (when it's too late).
* Audit this method to ensure that a client can only mutate metadata it has been given permission by the MDS to modify (i.e. what caps has it been issued). We should issue some kind of debug warning (level `-1`) when the MDS drops a cap update as that's **really** not good. Probably some kind of cluster warning too.
* Create tests which send synthetic cap updates for a client to the MDS which should trip on your new permission checks based on the issued caps.

Some other tests to consider: two users on a root_squashed mount, one root (uid=0) and other (uid=1?), both editing a file. The non-root user obtained full (all rights) caps first on the file. What should happen if the root uid tries to modify it on that client mount?

Reply:
I have created a [PR#48027](https://github.com/ceph/ceph/pull/48027) to make the client check the access in client side before buffering the changes locally or won't do a sync request for other operations. And no need to set the `caller_uid/gid` when doing the cap update.

Reply:
The [PR#48027](https://github.com/ceph/ceph/pull/48027) couldn't fix for the case when old clients. So we still need a fix in MDS side to make it work for that @ajarr, as you did here.

Reply:
W.r.t. old clients, the MDS does not error out an fsync() call -- that needs to be fixed too.

Reply:
> @batrick suggested to disallow old clients from mounting altogether rather than trying to fixup the (broken) behavior (for which we do not have optimal solutions).

Yeah, this sound more reasonable.